### PR TITLE
Formatting, Refactor, and fix Python backend import loop

### PR DESCRIPTION
**Pull request by pi314ever**
# What does this PR do?

Minor changes:
- Proper formatting with `cargo fmt` and `ruff format`
- Refactor `warmup_hpu` to not include rand dependency and use `max_batch_token` properly

Bug fix:
- Resolved import loop in Python where `Model` was incorrectly imported in `default_model.py`.

### CODE REVIEW

**Comment by yao-matrix** [comment]
@pi314ever , I don't prefer to add reformatting things into this PR, since this PR is mainly to add multi-backend design, not the reformat TEI coding-style, mixing things in one PR not comply w/ "one PR, one purpose" rule. if you found there is bug need to fix which is introduced by multi-backend design, pls append a bug-fixing PR to this.

You can submit another reformatting PR to tei repo once this multi-backend PR merged.

File: Cargo.toml [hunk 1]
```diff
@@ -29,8 +29,6 @@ tracing = "0.1"
 serde = { version = "1.0", features = ["serde_derive"] }
 serde_json = "1.0"
 thiserror = "1.0"
-rand = "0.8"
-
```
Comment (kaixuanliu) [inline Cargo.toml:33]: Let's keep it.

File: backends/Cargo.toml [hunk 2]
```diff
@@ -15,7 +15,6 @@ text-embeddings-backend-candle = { path = "candle", optional = true }
 text-embeddings-backend-ort = { path = "ort", optional = true }
 tokio = { workspace = true }
 tracing = { workspace = true }
-rand = { workspace = true }
```
Comment (kaixuanliu) [inline backends/Cargo.toml:18]: Keep it is OK

File: backends/python/server/requirements.txt [hunk 3]
```diff
@@ -40,4 +40,4 @@ typer==0.6.1 ; python_version >= "3.9" and python_version < "3.13"
 typing-extensions==4.7.1 ; python_version >= "3.9" and python_version < "3.13"
 urllib3==2.0.4 ; python_version >= "3.9" and python_version < "3.13"
 win32-setctime==1.1.0 ; python_version >= "3.9" and python_version < "3.13" and sys_platform == "win32"
-wrapt==1.15.0 ; python_version >= "3.9" and python_version < "3.13"
```
Comment (kaixuanliu) [inline backends/python/server/requirements.txt:43]: LGTM

File: backends/python/server/text_embeddings_server/models/__init__.py [hunk 4]
```diff
@@ -18,15 +18,14 @@
...
     from text_embeddings_server.models.flash_bert import FlashBert
+
+    __all__.append("FlashBert")
 except ImportError as e:
     logger.warning(f"Could not import Flash Attention enabled models: {e}")
     FLASH_ATTENTION = False
 
-if FLASH_ATTENTION:
-    __all__.append(FlashBert)
-
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/models/__init__.py:27]: Let's keep it.

File: backends/python/server/text_embeddings_server/models/__init__.py [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
+import torch
+from loguru import logger
 from transformers import AutoConfig
 from transformers.models.bert import BertConfig
 
-from text_embeddings_server.models.model import Model
 from text_embeddings_server.models.default_model import DefaultModel
+from text_embeddings_server.models.model import Model
 from text_embeddings_server.utils.device import get_device, use_ipex
+
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/models/__init__.py:12]: Let's keep it

File: backends/python/server/text_embeddings_server/models/__init__.py [hunk 6]
```diff
@@ -47,15 +46,18 @@ def get_model(model_path: Path, dtype: Optional[str]) :
...
         return DefaultModel(model_path, device, datatype)
     else:
         if device.type == "hpu":
             from habana_frameworks.torch.hpu import wrap_in_hpu_graph
-            from optimum.habana.transformers.modeling_utils import adapt_transformers_to_gaudi
+            from optimum.habana.transformers.modeling_utils import (
+                adapt_transformers_to_gaudi,
+            )
+
             adapt_transformers_to_gaudi()
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/models/__init__.py:61]: LGTM

File: backends/python/server/text_embeddings_server/models/default_model.py [hunk 7]
```diff
@@ -5,10 +5,10 @@
 from typing import Type, List
 from transformers import AutoModel
 from opentelemetry import trace
-from loguru import logger
 
-from text_embeddings_server.models import Model
+from text_embeddings_server.models.model import Model
 from text_embeddings_server.models.types import PaddedBatch, Embedding
+
 tracer = trace.get_tracer(__name__)
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/models/default_model.py:12]: LGTM

File: backends/python/server/text_embeddings_server/models/flash_bert.py [hunk 8]
```diff
@@ -66,14 +69,15 @@ def forward(self, hidden_states, residual=None):
...
             normed_hidden_states = ipex.llm.functional.add_layer_norm(
                 residual,
                 hidden_states,
                 self.weight,
                 self.bias,
                 self.variance_epsilon,
-                residual is not None
-             )
+                residual is not None,
+            )
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/models/flash_bert.py:81]: LGTM

File: backends/python/server/text_embeddings_server/models/flash_bert.py [hunk 9]
```diff
@@ -269,6 +273,7 @@ def __init__(self, model_path: Path, device: torch.device, dtype: torch.dtype):
             model = FlashBertModel(f, device, dtype, config)
         if device.type == "hpu":
             from habana_frameworks.torch.hpu import wrap_in_hpu_graph
+
             model = wrap_in_hpu_graph(model, disable_tensor_cache=False)
         self.hidden_size = config.hidden_size
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/models/flash_bert.py:279]: LGTM

File: backends/python/server/text_embeddings_server/models/model.py [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
-from typing import List, TypeVar, Type
+from typing import Generic, List, TypeVar, Type
 
 from text_embeddings_server.models.types import Batch, Embedding
 
 B = TypeVar("B", bound=Batch)
 
 
-class Model(ABC):
+class Model(ABC, Generic[B]):
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/models/model.py:11]: Let's keep it. Do not use this.

File: backends/python/server/text_embeddings_server/server.py [hunk 11]
```diff
@@ -8,7 +8,7 @@
 from pathlib import Path
 from typing import Optional
 
-from text_embeddings_server.models import Model, get_model
+from text_embeddings_server.models import get_model, Model
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/server.py:11]: LGTM

File: backends/python/server/text_embeddings_server/utils/device.py [hunk 12]
```diff
@@ -27,6 +32,7 @@ def get_major_and_minor_from_version(full_version):
         return False
     return True
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/utils/device.py:34]: Above is OK.

File: backends/python/server/text_embeddings_server/utils/device.py [hunk 13]
```diff
@@ -35,22 +41,25 @@ def is_hpu() -> bool:
         is_hpu_available = False
     return is_hpu_available
 
+
 def use_ipex() -> bool:
     value = os.environ.get("USE_IPEX", "True").lower()
-    return (value in ["true", "1"] and _is_ipex_available())
+    return value in ["true", "1"] and _is_ipex_available()
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/utils/device.py:47]: Add `()` is better, do not change it

File: backends/python/server/text_embeddings_server/utils/device.py [hunk 14]
```diff
@@ -35,22 +41,25 @@ def is_hpu() -> bool:
...
         import habana_frameworks.torch.core as htcore
-        if hasattr(torch, "hpu") and torch.hpu.is_available(): # type: ignore
+
+        if hasattr(torch, "hpu") and torch.hpu.is_available():  # type: ignore
             device = torch.device("hpu")
     elif use_ipex():
         import intel_extension_for_pytorch as ipex
+
         if hasattr(torch, "xpu") and torch.xpu.is_available():
             device = torch.device("xpu")
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/utils/device.py:63]: LGTM

File: backends/python/server/text_embeddings_server/utils/flash_attn.py [hunk 15]
```diff
@@ -120,13 +134,36 @@ def attention(q, k, v, out, cu_seqlens, max_s, softmax_scale, is_causal=False):
...
+                k,
+                v,
+                out,
+                cu_seqlens,
+                cu_seqlens,
+                max_s,
+                max_s,
+                softmax_scale,
+                is_causal=False,
+            )
```
Comment (kaixuanliu) [inline backends/python/server/text_embeddings_server/utils/flash_attn.py:167]: LGTM

File: backends/src/lib.rs [hunk 16]
```diff
@@ -41,10 +40,11 @@ fn powers_of_two(max_value: usize) -> Vec<usize> {
...
     match Command::new("hl-smi")
-        .args(&["-Q", "name", "-f", "csv"])
-        .output() {
-            Ok(output) => output.status.success(),
-            Err(_) => false,
+        .args(["-Q", "name", "-f", "csv"])
+        .output()
+    {
+        Ok(output) => output.status.success(),
+        Err(_) => false,
```
Comment (kaixuanliu) [inline backends/src/lib.rs:47]: Do not change it

**Comment by pi314ever** [comment]
@yao-matrix I agree with one PR one purpose, but I believe that the code should maintain formatting guidelines already followed by the upstream repository. All of the changes in this PR deals with code we will add in support of multi-backend. 

The changes to exclude the rand package is to resolve inconsistency in how `max_batch_tokens` is interpreted. In the upstream warmup function, it defines the maximum number of tokens allowed in each batch. However, we were using it as the maximum token ID. The removal of the rand package makes it more consistent with their version of warmup. 

The bug fix was simply an import loop that occurs in default model, which is a single line fix.

File: backends/src/lib.rs [hunk 17]
```diff
@@ -99,20 +99,19 @@ impl Backend {
     #[instrument(skip(self))]
     pub async fn warmup_hpu(
         &self,
-        mut max_input_length: usize,
-        max_token: usize,
```
Comment (kaixuanliu) [inline backends/src/lib.rs:103]: `max_input_length` may be changed in this function.

File: backends/src/lib.rs [hunk 18]
```diff
@@ -130,9 +129,24 @@ impl Backend {
...
+                BackendError::Start(
+                    format!(
+                        "Largest warmup batch ({} tokens/batch * {} batches = {} tokens) exceeds maximum batch token size ({}). Please reduce the max warmup sequence length or max warup batch size.",
+                        max_warmup_length,
+                        max_batch_size,
+                        max_warmup_length * max_batch_size,
+                        max_batch_token
+                    )
+                )
+            );
```
Comment (kaixuanliu) [inline backends/src/lib.rs:143]: LGTM. Add this check is better.

File: backends/src/lib.rs [hunk 19]
```diff
@@ -130,9 +129,24 @@ impl Backend {
...
+        }
 
-        max_input_length = std::cmp::min(max_input_length, max_warmup_length);
-        let mut seq_lengths: Vec<usize> = (seq_bucket_size..max_input_length+1).step_by(seq_bucket_size as usize).collect();
+        let max_input_length = std::cmp::min(max_input_length, max_warmup_length);
+        let mut seq_lengths: Vec<usize> = (seq_bucket_size..max_input_length + 1)
+            .step_by(seq_bucket_size)
+            .collect();
```
Comment (kaixuanliu) [inline backends/src/lib.rs:149]: LGTM

File: backends/src/lib.rs [hunk 20]
```diff
@@ -146,7 +160,7 @@ impl Backend {
             }
         }
         for shape in shapes.iter() {
-            let batch = self.create_warmup_batch(*shape, max_token as u32);
+            let batch = self.create_warmup_batch(*shape);
```
Comment (kaixuanliu) [inline backends/src/lib.rs:163]: Let's keep it

File: backends/src/lib.rs [hunk 21]
```diff
@@ -157,20 +171,16 @@ impl Backend {
...
         let mut batched_input_ids = Vec::new();
         let mut batched_token_type_ids = Vec::new();
         let mut batched_position_ids = Vec::new();
         let mut cumulative_seq_lengths = Vec::with_capacity(batch_size as usize + 1);
         let mut pooled_indices = Vec::with_capacity(batch_size as usize);
         cumulative_seq_lengths.push(0);
-        let input_ids: Vec<u32> = (0..length).map(|_| rand::thread_rng().gen_range(0..max_token)).collect();
-        let token_type_ids: Vec<u32> = vec![0; length as usize];
+        let input_ids = vec![0; length as usize];
+        let token_type_ids = vec![0; length as usize];
```
Comment (kaixuanliu) [inline backends/src/lib.rs:183]: Let's keep it.

File: backends/src/lib.rs [hunk 22]
```diff
@@ -200,7 +210,9 @@ impl Backend {
         max_batch_requests: Option<usize>,
     ) -> Result<(), BackendError> {
         if is_hpu() {
-            return self.warmup_hpu(max_input_length, max_batch_tokens, max_batch_requests).await;
+            return self
+                .warmup_hpu(max_input_length, max_batch_tokens, max_batch_requests)
+                .await;
         }
```
Comment (kaixuanliu) [inline backends/src/lib.rs:216]: LGTM

**Comment by kaixuanliu** [comment]
@pi314ever , Hi, I add some comments. Some code style re-format is good. It looks better. However, I have some points of suggestions:
1. It is not recommended to use too heavy extension tools to check code style, some changes like this is not necessary: [L11](https://github.com/yuanwu2017/text-embeddings-inference/pull/7/files#diff-3b2eb5c3c3fcfd7fbaa54c191969dd169959aeb203da43dd22634a1da88909ddR11)
2. In this PR, it is better not change original code if not necessary, we can submit new PR to fix it. Example:[R22](https://github.com/yuanwu2017/text-embeddings-inference/pull/7/files#diff-78a9ad75109c914e10c838195a64cff1395c6729ba0680602db904ec5ae8734eR22)
3. As for warmup logic, here I want to use random input instead of all zeros. And I need to set a reasonable value(not too big) here for `rand::thread_rng().gen_range` API. It is not a big problem, we can get feedback from reviewers.

**Comment by pi314ever** [comment]
Closing this PR in favor of #8 as the bug is a non-issue.
